cpp/test/sg/fil_test.cu

```diff
@@ -50,6 +53,15 @@ struct FilTestParams {
   int depth       = 8;
   int num_trees   = 50;
   float leaf_prob = 0.05;
+  // below, categorical nodes means categorical branch nodes
+  // probability that a node is categorical (given that its feature is categorical)
+  float node_categorical_prob = 0.0;
+  // probability that a feature is categorical (given that the node is also categorical)
+  float feature_categorical_prob = 0.0;
```

Comment:
This set of two conditional probabilities sounds like an endless loop. I guess the probability of a feature being categorical shouldn't be conditional.

Reply:
thank you, yes!